// qq-connect/lib/qzone.js
var https = require('https'),
	querystring = require('querystring');
var config = require('./config');
var get_user_info_url = 'https://graph.qq.com/user/get_user_info',
	qzone_host = 'graph.qq.com',
	add_share_path = '/share/add_share';
var format_common_url = function(params) {
	var common_url = '?access_token=' + params.access_token + 
		'&oauth_consumer_key=' + config.appid + 
		'&openid=' + params.openid;
	return common_url;
};
var get_user_info = function(params, cb, er) {
	var get_user_info_url2 = get_user_info_url + format_common_url(params);
	https.get(get_user_info_url2, function(res) {
		var result = '';
		res.on('data', function(d) {
	    result += d;
	  });
	  res.on('end', function() {
	  	cb(JSON.parse(result));
	  });
	  res.on('error', function(e) {
	  	er(e);
	  });
	});
};
var add_share = function(params, cb, er) {
	var data = querystring.stringify({
    access_token: params.access_token,
    oauth_consumer_key: config.appid,
    openid: params.openid,
    title: params.title,
    url: params.url,
    site: params.site,
    fromurl: params.fromurl
  });
	var options = {
	    host: 'proxy.tencent.com',
	    port: 8080,
	    path: 'https://graph.qq.com/share/add_share',
	    method: 'POST',
	    headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Content-Length': data.length
	    }
	};
	var req = require('http').request(options, function(res) {
	    res.setEncoding('utf8');
	    var result = null;
	    res.on('data', function (chunk) {
	      console.log("body: " + chunk);
	      result = chunk;
	    });
	    res.on('end', function() {
	    	cb(result)
	    });
	    res.on('error', function(e) {
	    	er(e);
	    });
	});
	req.on('error', function(e) {
		er(e);
	});
	req.write(data);
	req.end();
};
exports = module.exports = {
	get_user_info: get_user_info,
	add_share: add_share
};
